File: app/java/src/main/java/com/example/production/infrastructure/web/controller/ItemController.java
package com.example.production.infrastructure.web.controller;

import com.example.production.application.port.in.ItemUseCase;
import com.example.production.application.port.in.command.CreateItemCommand;
import com.example.production.application.port.in.command.UpdateItemCommand;
import com.example.production.domain.model.item.Item;
import com.example.production.domain.model.item.ItemCategory;
import com.example.production.infrastructure.web.dto.CreateItemRequest;
import com.example.production.infrastructure.web.dto.ItemResponse;
import com.example.production.infrastructure.web.dto.UpdateItemRequest;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.Parameter;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * 品目 Controller（Input Adapter）
 */
@RestController
@RequestMapping("/api/items")
@Tag(name = "items", description = "品目マスタ API")
@RequiredArgsConstructor
public class ItemController {

    private final ItemUseCase itemUseCase;

    @GetMapping
    @Operation(
            summary = "品目一覧の取得",
            description = "すべての品目を取得します。category クエリパラメータでフィルタリング可能です。"
    )
    public ResponseEntity<List<ItemResponse>> getAllItems(
            @Parameter(description = "品目区分でフィルタリング")
            @RequestParam(required = false) ItemCategory category) {

        List<Item> items;
        if (category != null) {
            items = itemUseCase.getItemsByCategory(category);
        } else {
            items = itemUseCase.getAllItems();
        }

        return ResponseEntity.ok(items.stream()
                .map(ItemResponse::from)
                .toList());
    }

    @GetMapping("/category/{category}")
    @Operation(
            summary = "カテゴリ別品目一覧の取得",
            description = "指定したカテゴリの品目を取得します"
    )
    public ResponseEntity<List<ItemResponse>> getItemsByCategory(
            @Parameter(description = "品目区分", required = true)
            @PathVariable ItemCategory category) {

        List<Item> items = itemUseCase.getItemsByCategory(category);
        return ResponseEntity.ok(items.stream()
                .map(ItemResponse::from)
                .toList());
    }

    @GetMapping("/{itemCode}")
    @Operation(summary = "品目の取得")
    @ApiResponse(responseCode = "200", description = "成功")
    @ApiResponse(responseCode = "404", description = "品目が見つからない")
    public ResponseEntity<ItemResponse> getItemByCode(
            @Parameter(description = "品目コード", required = true)
            @PathVariable String itemCode) {

        Item item = itemUseCase.getItemByCode(itemCode);
        return ResponseEntity.ok(ItemResponse.from(item));
    }

    @PostMapping
    @Operation(summary = "品目の登録")
    @ApiResponse(responseCode = "201", description = "登録成功")
    @ApiResponse(responseCode = "409", description = "品目コード重複")
    public ResponseEntity<ItemResponse> createItem(
            @Valid @RequestBody CreateItemRequest request) {

        CreateItemCommand command = CreateItemCommand.builder()
                .itemCode(request.getItemCode())
                .itemName(request.getItemName())
                .category(request.getCategory())
                .unitCode(request.getUnitCode())
                .leadTime(request.getLeadTime())
                .safetyLeadTime(request.getSafetyLeadTime())
                .safetyStock(request.getSafetyStock())
                .yieldRate(request.getYieldRate())
                .minLotSize(request.getMinLotSize())
                .lotIncrement(request.getLotIncrement())
                .maxLotSize(request.getMaxLotSize())
                .shelfLife(request.getShelfLife())
                .build();

        Item item = itemUseCase.createItem(command);
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(ItemResponse.from(item));
    }

    @PutMapping("/{itemCode}")
    @Operation(summary = "品目の更新")
    public ResponseEntity<ItemResponse> updateItem(
            @PathVariable String itemCode,
            @Valid @RequestBody UpdateItemRequest request) {

        UpdateItemCommand command = UpdateItemCommand.builder()
                .itemCode(itemCode)
                .itemName(request.getItemName())
                .category(request.getCategory())
                .leadTime(request.getLeadTime())
                .safetyLeadTime(request.getSafetyLeadTime())
                .safetyStock(request.getSafetyStock())
                .yieldRate(request.getYieldRate())
                .minLotSize(request.getMinLotSize())
                .lotIncrement(request.getLotIncrement())
                .maxLotSize(request.getMaxLotSize())
                .shelfLife(request.getShelfLife())
                .build();

        Item item = itemUseCase.updateItem(command);
        return ResponseEntity.ok(ItemResponse.from(item));
    }

    @DeleteMapping("/{itemCode}")
    @Operation(summary = "品目の削除")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteItem(@PathVariable String itemCode) {
        itemUseCase.deleteItem(itemCode);
    }
}

File: app/csharp/src/ProductionManagement.Infrastructure/Rest/Controllers/ItemController.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using ProductionManagement.Application.Port.In;
using ProductionManagement.Domain.Models.Item;
using ProductionManagement.Infrastructure.Rest.Dto;

namespace ProductionManagement.Infrastructure.Rest.Controllers;

/// <summary>
/// 品目 Controller（Input Adapter）
/// 既存の IItemUseCase を Input Adapter として呼び出す
/// </summary>
[ApiController]
[Route("api/items")]
[Tags("Items")]
public class ItemController : ControllerBase
{
    private readonly IItemUseCase _itemUseCase;

    public ItemController(IItemUseCase itemUseCase)
    {
        _itemUseCase = itemUseCase;
    }

    /// <summary>
    /// 品目一覧の取得
    /// </summary>
    /// <param name="category">品目区分でフィルタリング</param>
    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<ItemResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAllItems([FromQuery] ItemCategory? category)
    {
        IReadOnlyList<Item> items;
        if (category.HasValue)
        {
            items = await _itemUseCase.GetItemsByCategoryAsync(category.Value);
        }
        else
        {
            items = await _itemUseCase.GetAllItemsAsync();
        }

        return Ok(items.Select(ItemResponse.From).ToList());
    }

    /// <summary>
    /// 品目の取得
    /// </summary>
    /// <param name="itemCode">品目コード</param>
    [HttpGet("{itemCode}")]
    [ProducesResponseType(typeof(ItemResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetItemByCode(string itemCode)
    {
        var item = await _itemUseCase.GetItemByCodeAsync(itemCode);
        return Ok(ItemResponse.From(item));
    }

    /// <summary>
    /// 品目の登録
    /// </summary>
    [HttpPost]
    [ProducesResponseType(typeof(ItemResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    public async Task<IActionResult> CreateItem([FromBody] CreateItemRequest request)
    {
        var command = new CreateItemCommand(
            ItemCode: request.ItemCode,
            ItemName: request.ItemName,
            Category: Enum.Parse<ItemCategory>(request.Category),
            UnitCode: request.UnitCode,
            LeadTime: request.LeadTime,
            SafetyLeadTime: request.SafetyLeadTime,
            SafetyStock: request.SafetyStock,
            YieldRate: request.YieldRate,
            MinLotSize: request.MinLotSize,
            LotIncrement: request.LotIncrement,
            MaxLotSize: request.MaxLotSize,
            ShelfLife: request.ShelfLife
        );

        var item = await _itemUseCase.CreateItemAsync(command);
        return CreatedAtAction(
            nameof(GetItemByCode),
            new { itemCode = item.ItemCode },
            ItemResponse.From(item)
        );
    }

    /// <summary>
    /// 品目の更新
    /// </summary>
    [HttpPut("{itemCode}")]
    [ProducesResponseType(typeof(ItemResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> UpdateItem(
        string itemCode,
        [FromBody] UpdateItemRequest request)
    {
        var command = new UpdateItemCommand(
            ItemCode: itemCode,
            ItemName: request.ItemName,
            Category: request.Category is not null
                ? Enum.Parse<ItemCategory>(request.Category)
                : null,
            UnitCode: request.UnitCode,
            LeadTime: request.LeadTime,
            SafetyLeadTime: request.SafetyLeadTime,
            SafetyStock: request.SafetyStock,
            YieldRate: request.YieldRate,
            MinLotSize: request.MinLotSize,
            LotIncrement: request.LotIncrement,
            MaxLotSize: request.MaxLotSize,
            ShelfLife: request.ShelfLife
        );

        var item = await _itemUseCase.UpdateItemAsync(command);
        return Ok(ItemResponse.From(item));
    }

    /// <summary>
    /// 品目の削除
    /// </summary>
    [HttpDelete("{itemCode}")]
    [ProducesResponseType(StatusCodes.Status204NoContent)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> DeleteItem(string itemCode)
    {
        await _itemUseCase.DeleteItemAsync(itemCode);
        return NoContent();
    }
}
